Shipwright often fails when build starts from an empty docker local images repository

Hello

We are experiencing some weird behaviors with shipwright, when building from a host which does not have any docker images already available on the local repo.

```
ecpff/ecp_bamboo_haproxy | Step 0 : FROM ecpff/ecp_base:85c29768586c
ecpff/ecp_bamboo_haproxy | [STATUS] : Pulling repository ecpff/ecp_base
ecpff/ecp_bamboo_haproxy | [ERROR] Error: image ecpff/ecp_base:85c29768586c not found

ecpff/ecp_base | Step 0 : FROM ecpff/ecp_scratch:85c29768586c
ecpff/ecp_base |  ---> 1137614cd2b3
ecpff/ecp_base | Step 1 : "blabla"
ecpff/ecp_base |  ---> Running in ea17fcd2c6d8
ecpff/ecp_base |  ---> d17262206bbb
ecpff/ecp_base | Removing intermediate container fc3c1aecb76e
ecpff/ecp_base | Successfully built 75b1a04a07f6
Traceback (most recent call last):
  File "/usr/local/bin/shipwright", line 9, in <module>
    load_entry_point('shipwright==0.2.2', 'console_scripts', 'shipwright')()
  File "/usr/local/lib/python2.7/dist-packages/shipwright/cli.py", line 193, in main
    for event in command(*args):
  File "/usr/local/lib/python2.7/dist-packages/shipwright/__init__.py", line 115, in build_tree
    branch
  File "/usr/local/lib/python2.7/dist-packages/shipwright/docker.py", line 54, in tag_containers
    force=True
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 978, in tag
    self._raise_for_status(res)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 94, in _raise_for_status
    raise errors.APIError(e, response, explanation=explanation)
docker.errors.APIError: 404 Client Error: Not Found ("no such id: ecpff/ecp_bamboo_haproxy:85c29768586c")
```

If no image is already available (build from scratch), then:
- it loops on images, unable to get the parent image [normal]
- when it gets to the root image, it builds it [seems also normal]
- then it often fails with a "no such id" error...

If a restart shipwright (several times), it goes further and further until the end of the build process.


I don't have any debug information to provide right now, but if needed just ask and i can generate them.

Sorry for the delay!
Here are the requested details


Folder structure & FROM field of each docker file
```
.
.shipwright.json
├── ecp_bamboo_haproxy      -> FROM ecpff/ecp_base
├── ecp_base                -> FROM ecpff/ecp_scratch
├── ecp_ffaas_backoffice    -> FROM ecpff/ecp_scm
├── ecp_jenkins_base        -> FROM ecpff/ecp_scm
├── ecp_scm                 -> FROM ecpff/ecp_bas
├── ecp_scratch             -> FROM scratch
```

.shipwright.json content
```json
{
  "version": 1.0,
  "namespace": "ecpff"
}
```

Docker local images repository before executing shipwright
```
$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
```

Docker local images repository after executing shipwright
```
$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
ecpff/ecp_scm       8f11650a63e5        d692f8f7c301        13 minutes ago      153 MB
ecpff/ecp_base      8f11650a63e5        e603056dfa86        15 minutes ago      66.02 MB
ecpff/ecp_scratch   8f11650a63e5        991470ffd92e        16 minutes ago      54.52 MB
scratch             latest              511136ea3c5a        21 months ago       0 B
```

For reminder, here is the *stripped* shipwright build trace
```
$ shipwright
ecpff/ecp_bamboo_haproxy | Step 0 : FROM ecpff/ecp_base:8f11650a63e5
ecpff/ecp_bamboo_haproxy | [STATUS] : Pulling repository ecpff/ecp_base
ecpff/ecp_bamboo_haproxy | [ERROR] Error: image ecpff/ecp_base:8f11650a63e5 not found

ecpff/ecp_ffaas_backoffice | Step 0 : FROM ecpff/ecp_scm:8f11650a63e5
ecpff/ecp_ffaas_backoffice | [STATUS] : Pulling repository ecpff/ecp_scm
ecpff/ecp_ffaas_backoffice | [ERROR] Tag 8f11650a63e5 not found in repository ecpff/ecp_scm

ecpff/ecp_jenkins_base | Step 0 : FROM ecpff/ecp_scm:8f11650a63e5
ecpff/ecp_jenkins_base | [STATUS] : Pulling repository ecpff/ecp_scm
ecpff/ecp_jenkins_base | [ERROR] Tag 8f11650a63e5 not found in repository ecpff/ecp_scm

ecpff/ecp_scratch | Step 0 : FROM scratch
ecpff/ecp_scratch | [STATUS] : Pulling repository scratch
ecpff/ecp_scratch | [STATUS] 511136ea3c5a: Pulling image (latest) from scratch
ecpff/ecp_scratch | [STATUS] 511136ea3c5a: Pulling image (latest) from scratch, endpoint: https://registry-1.docker.io/v1/
ecpff/ecp_scratch | [STATUS] 511136ea3c5a: Pulling dependent layers
ecpff/ecp_scratch | [STATUS] 511136ea3c5a: Pulling metadata
ecpff/ecp_scratch | [STATUS] 511136ea3c5a: Pulling fs layer
ecpff/ecp_scratch | [STATUS] 511136ea3c5a: Downloading
ecpff/ecp_scratch | [STATUS] 511136ea3c5a: Download complete
ecpff/ecp_scratch | [STATUS] 511136ea3c5a: Download complete
ecpff/ecp_scratch | [STATUS] : Status: Downloaded newer image for scratch:latest
ecpff/ecp_scratch |  ---> 511136ea3c5a
ecpff/ecp_scratch | Step 1 : ADD rootfs.tar.xz /
ecpff/ecp_scratch |  ---> 991470ffd92e
ecpff/ecp_scratch | Removing intermediate container 96fa1f7d12fe
ecpff/ecp_scratch | Successfully built 991470ffd92e
ecpff/ecp_base | Step 0 : FROM ecpff/ecp_scratch:8f11650a63e5
ecpff/ecp_base |  ---> 991470ffd92e
ecpff/ecp_base | Step 1 : MAINTAINER user <user@user.com>
ecpff/ecp_base |  ---> Running in 2c0ab2847529
ecpff/ecp_base |  ---> db3d8803e2d1
ecpff/ecp_base | Removing intermediate container 2c0ab2847529
...
ecpff/ecp_base | Step 15 : ENV HOME /home/ecp
ecpff/ecp_base |  ---> Running in 1217fcb96e5e
ecpff/ecp_base |  ---> e603056dfa86
ecpff/ecp_base | Removing intermediate container 1217fcb96e5e
ecpff/ecp_base | Successfully built e603056dfa86
ecpff/ecp_scm | Step 0 : FROM ecpff/ecp_base:8f11650a63e5
ecpff/ecp_scm |  ---> e603056dfa86
ecpff/ecp_scm | Step 1 : MAINTAINER user <user@user.com>
ecpff/ecp_scm |  ---> Running in 2bdd3c79174e
ecpff/ecp_scm |  ---> efd1167fdb51
ecpff/ecp_scm | Removing intermediate container 2bdd3c79174e
...
ecpff/ecp_scm | Step 7 : RUN git config --global user.email "ecp@no-answer"
ecpff/ecp_scm |  ---> Running in 8dcd88e63400
ecpff/ecp_scm |  ---> d692f8f7c301
ecpff/ecp_scm | Removing intermediate container 8dcd88e63400
ecpff/ecp_scm | Successfully built d692f8f7c301
Traceback (most recent call last):
  File "/usr/local/bin/shipwright", line 9, in <module>
    load_entry_point('shipwright==0.2.2', 'console_scripts', 'shipwright')()
  File "/usr/local/lib/python2.7/dist-packages/shipwright/cli.py", line 193, in main
    for event in command(*args):
  File "/usr/local/lib/python2.7/dist-packages/shipwright/__init__.py", line 115, in build_tree
    branch
  File "/usr/local/lib/python2.7/dist-packages/shipwright/docker.py", line 54, in tag_containers
    force=True
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 934, in tag
    self._raise_for_status(res)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 119, in _raise_for_status
    raise errors.APIError(e, response, explanation=explanation)
docker.errors.APIError: 404 Client Error: Not Found ("no such id: ecpff/ecp_bamboo_haproxy:8f11650a63e5")
```

--- a/shipwright/__init__.py
+++ b/shipwright/__init__.py
@@ -72,7 +72,6 @@ class Shipwright(object):
 
 
   def build(self, specifiers):
-    
     tree = dependencies.eval(specifiers, self.targets())
     return self.build_tree(tree)
 

--- a/shipwright/container.py
+++ b/shipwright/container.py
@@ -50,7 +50,7 @@ def containers(name_func, path):
   ]
 
 
-#namespace -> path -> Container(name, path, parent)
+# (path -> name) -> path -> Container(name, path, parent)
 def container_from_path(name_func, path):
   """
   Given a path to a Dockerfile parse the file and return 

--- a/shipwright/dependencies.py
+++ b/shipwright/dependencies.py
@@ -55,7 +55,6 @@ def eval(specifiers, targets):
 def needs_building(tree):
   """
   """
-  
   gen = breadth_first_iter(tree)
   root = next(gen) # skip root
   loc = next(gen)
@@ -67,9 +66,8 @@ def needs_building(tree):
   while True:
     try:
       target  = loc.node()
-      
-      if target.current_rel > target.last_built_rel:
 
+      if target.current_rel > target.last_built_rel:
         # target has changes, it and all it's desendents need
         # to be rebuilt
         for modified_loc in breadth_first_iter(loc):
@@ -129,34 +127,23 @@ def make_tree(containers):
   tree =  zipper.zipper(root, is_branch, children, make_node)
   
   for c in containers:
-    t = c._replace(children=())
+
+    branch_children, root_children = split(is_child(c), tree.children())
+    t = c._replace(children=tuple(branch_children))
+
+    if branch_children:
+      tree = tree.edit(replace, tuple(root_children))
 
     loc = tree.find(fmap(is_target(t.parent)))
     if loc:
       tree = loc.insert(t).top()
-    else: 
-      tree = tree.edit(reroot, t)
+    else:
+      tree = tree.insert(t) 
 
   return tree
 
-def reroot(root, branch):
-  """
-  Adds the branch into the root, sweeping up children that
-  may have been inserted before it.
-  """
-  root_children = []
-  branch_children = []
-  for child in root.children:
-    if child.parent == branch.name:
-      branch_children.append(child)
-    else:
-      root_children.append(child)
-
-  branch = branch._replace(children=tuple(branch_children))
-  root_children.append(branch)
-
-  return root._replace(children=tuple(root_children))
-
+def replace(node, children):
+  return node._replace(children=children)
 
 def children(item):
   return item.children
@@ -211,7 +198,7 @@ def is_target(name, target):
 @curry
 def is_child(parent, target):
   if not isinstance(target, Root):
-    return target.parent == parent
+    return target.parent == parent.name
 
 # (a -> b) -> Loc a -> b
 @curry
@@ -226,6 +213,26 @@ def lineage(loc):
     results.append(node)
     loc = loc.up()
   return results
+
+
+# (a -> Bool) -> [a] ->[a], [a]
+@curry
+def split(f, children):
+  """
+  Given a function that returns true or false and a list. Return
+  a two lists all items f(child) == True is in list 1 and 
+  all items not in the list are in list 2.
+
+  """
+
+  l1 = []
+  l2 = []
+  for child in children:
+    if f(child):
+      l1.append(child)
+    else:
+      l2.append(child)
+  return l1, l2
 
 
 # Loc -> [Target]

--- a/shipwright/version.py
+++ b/shipwright/version.py
@@ -1,1 +1,1 @@
-version = "0.2.2"
+version = "0.2.3"

